com.unity.netcode.gameobjects/Prototyping/Interpolator/BufferedLinearInterpolator.cs

```diff
@@ -0,0 +1,232 @@
+            {
+                // buffer is sorted so older (smaller) time values are at the end.
+                for (int i = m_Buffer.Count - 1; i >= 0; i--) // todo stretch: consume ahead if we see we're missing values
+                {
+                    var bufferedValue = m_Buffer[i];
+                    // Consume when ready. This can consume multiple times
+                    if (FixedOrTime(bufferedValue.timeSent) <= ServerTimeBeingHandledForBuffering)
+                    {
+                        if (m_LifetimeConsumedCount == 0)
+                        {
+                            m_StartTimeConsumed = bufferedValue.timeSent;
+                            m_InterpStartValue = bufferedValue.item;
+                        }
+                        else if (consumedCount == 0)
+                        {
+                            m_StartTimeConsumed = m_EndTimeConsumed;
+                            m_InterpStartValue = m_InterpEndValue;
+                        }
+
+                        m_EndTimeConsumed = bufferedValue.timeSent;
+                        m_InterpEndValue = bufferedValue.item;
+                        m_Buffer.RemoveAt(i);
+                        consumedCount++;
+                        m_LifetimeConsumedCount++;
+                    }
+                }
+            }
+        }
+
+        public T Update(float deltaTime)
+        {
+            TryConsumeFromBuffer();
+
+            if (m_LifetimeConsumedCount == 0 && m_Buffer.Count == 0)
+            {
+                throw new InvalidOperationException("trying to update interpolator when no data has been added to it yet");
+            }
+
+            // Interpolation example to understand the math below
+            // 4   4.5      6   6.5
+            // |   |        |   |
+            // A   render   B   Server
+
+            if (m_LifetimeConsumedCount >= 1) // shouldn't interpolate between default values, let's wait to receive data first, should only interpolate between real measurements
+            {
+                double range = FixedOrTime(m_EndTimeConsumed) - FixedOrTime(m_StartTimeConsumed);
+                float t;
+                if (range == 0)
+                {
+                    t = 1;
+                }
+                else
+                {
+                    t = (float) ((RenderTime - FixedOrTime(m_StartTimeConsumed)) / range);
+                }
+
+                if (t > 3) // max extrapolation
+                {
+                    t = 1;
+                }
+
+                if (Debug.isDebugBuild)
+                {
+                    Debug.Assert(t >= 0, $"t must be bigger or equal than 0. range {range}, RenderTime {RenderTime}, Start time {FixedOrTime(m_StartTimeConsumed)}, end time {FixedOrTime(m_EndTimeConsumed)}");
+                }
+
+                SimpleInterpolator.AddMeasurement(Interpolate(m_InterpStartValue, m_InterpEndValue, t), default); // using simple interpolation so there's no jump
```

Comment:
I think I understand the idea of `BufferedLinearInterpolator` it buffers measurements and then feeds them into the simple interpolator. I'm confused though by how the `SimpleInterpolator` works.